coaster/sqlalchemy/roles.py

```diff
@@ -269,6 +270,112 @@ def copy(self):
     symmetric_difference_update = nary_op(collections.MutableSet.__ixor__)
 
 
+class LazyAssociationProxy(object):
+    """
+    Lazy set that acts as the association proxy of given lazy relationship.
+    This is only required in case the relationship has `lazy='dynamic'` defined.
+
+    Example usage:
+
+        class ChildDocument (db.Model):
+            property = db.Column()
+
+        class Document(db.Model):
+            pass
+
+        Document.collection_name = db.relationship(
+            ChildDocument,
+            lazy='dynamic,
+            ...
+        )
+
+        Document.properties = LazyAssociationProxy('collection_name', 'property')
+
+    """
+
+    def __init__(self, target_collection, attr):
+        self.target_collection = target_collection
+        self.attr = attr
+
+    def __get__(self, obj, cls):
+        if obj is None:
+            return self
+        return LazyAssociationProxyWrapper(obj, self.target_collection, self.attr, cls)
+
+
+class LazyAssociationProxyWrapper(collections.Set):
+    def __init__(self, obj, target_collection, attr, cls=None):
+        self.obj = obj
+        self.target_collection = target_collection
+        self.attr = attr
+        self.cls = cls
+
+        self._target_collection_obj = getattr(self.obj, self.target_collection)
+
+        # if _length is 0 by default, and then the calculated count is 0 as well,
+        # this cache variable will never get used. So setting it to None.
+        self._length = None
+
+    def __repr__(self):  # pragma: no cover
+        return "LazyAssociationProxy generator (%r, %r)" % (
+            self.target_collection,
+            self.attr,
+        )
+
+    def __getattr__(self, name):
+        return getattr(self.obj, name)
+
+    def _member_is_present(self, member):
+        if member is not None:
+            if isinstance(self._target_collection_obj, LazyAssociationProxy):
+                return member in getattr(self._target_collection_obj, self.attr)
+            else:
+                # A lazy sqlalchemy relationship
+                return self._target_collection_obj.session.query(
+                    self._target_collection_obj.filter_by(
+                        **{self.attr: member}
+                    ).exists()
+                ).scalar()
+        return False
+
+    def _contents(self):
+        # Return all available members
+        for target_obj in self._target_collection_obj:
+            yield getattr(target_obj, self.attr)
+
+    def __contains__(self, member):
+        return self._member_is_present(member)
+
+    __iter__ = _contents
+
+    def _len(self):
+        self._length = self._target_collection_obj.count()
+        return self._length
+
+    def __len__(self):
+        # check cache first and then calculate if needed
+        return self._len() if self._length is None else self._length
```

Comment:
It's not clear to me why you have methods that simply call other methods. Make the code inline.